### PR TITLE
[WIP] SARIMA Model (Time series - GSoC)

### PR DESCRIPTION
**Pull request by mjhajharia**
so before any comments, this is a very very very  rudimentary PR, right now i've only made it into a PR to be able to hear comments. My current plan for implementing sarima is:

- [ ] convert [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class
- [ ] add aesara and suitable pymc functionalities to make it a pymc distribution (a separate submodule as @fonnesbeck suggested)
- [ ] Add Kalman filter once we have a working sarima distribution in pymc3
- [ ] additional improvements // features like a subclass of arma, ar, ma, sarimax + ipython notebooks with tutorials

so right now, i'm in the part where i've created a rudimentary conversion of [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class, its not working yet, the function definitions etc are all over the place, ive just essentially translated stan code to numpy/pymc equivalent. i could definitely definitely use opinions/suggestions/detailed comments/feedback anything!!!

### CODE REVIEW

**Comment by mjhajharia** [comment]
@ColCarroll thanks for the inverse chi squared logp!!!!!!

**Comment by canyon289** [comment]
This is a quick 1 minute high level review. Documentation and general model looks great at a high level. Will review more deeply soon. Can a test be added to show functionality and give an example to be run? it can be something really simple for now just to start

Great work, this is very exciting

File: pymc3/distributions/arimastan.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
+import numpy as np
+
+import pymc3 as pm
+
+
+def invchi(x, chisquare):
+    return (1.0 / x ** 2) * chisquare.logp(1 / x)
+
+
+class Sarima:
```
Comment (fonnesbeck) [inline pymc3/distributions/arimastan.py]: SARIMA, to be consistent with AR, GARCH, etc.

**Comment by fonnesbeck** [comment]
ARIMA models are complex enough that they probably should not be in with the distributions, but rather, have a submodule of their own (just as GPs do).

**Comment by mjhajharia** [comment]
> ARIMA models are complex enough that they probably should not be in with the distributions, but rather, have a submodule of their own (just as GPs do).

makes sense yeah, like we could later take AR as a part of arima too, just arima(1,0,0) more or less ig. eventually i plan to have something like statsmodels sarimax - which has arima, sarima, sarimax, ar, ma

File: pymc3/distributions/arimastan.py (hunk 1 above)
Comment (mjhajharia) [reply pymc3/distributions/arimastan.py]: will do, thanks!

File: pymc3/distributions/arimastan.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            if self.prior_sma[i, 4] == 1:
+                self.stheta[i] = self.Theta0[i]
+            else:
+                self.stheta[i] = 2 * abs(self.Theta0[i]) - 1
+
+    # model
+
+    def prior_for_mu0(self, target, prior_for_mu0):
+        # prior mu
+        if self.prior_mu0[4] == 1:
```
Comment (canyon289) [inline pymc3/distributions/arimastan.py]: Is the idea to set prior distributions using integer arguments? In general there is a lot of code repetition here and wondering if we can simplify

**Comment by canyon289** [comment]
> so before any comments, this is a very very very rudimentary PR, right now i've only made it into a PR to be able to hear comments. My current plan for implementing sarima is:
> 
>     * [ ]  convert [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class
> 
>     * [ ]  add aesara and suitable pymc functionalities to make it a pymc distribution (a separate submodule as @fonnesbeck suggested)
> 
>     * [ ]  Add Kalman filter once we have a working sarima distribution in pymc3
> 
>     * [ ]  additional improvements // features like a subclass of arma, ar, ma, sarimax + ipython notebooks with tutorials
> 
> 
> so right now, i'm in the part where i've created a rudimentary conversion of [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class, its not working yet, the function definitions etc are all over the place, ive just essentially translated stan code to numpy/pymc equivalent. i could definitely definitely use opinions/suggestions/detailed comments/feedback anything!!!

Some comments
* This is a great draft PR. Thank you for sharing. If we ask a lot of questions its not because you did anything wrong, more so that this is the perfect place to ask precise questions and have great concrete discussion
* Suggestion: In this PR I would focus on the first two items in the checklist, and the follow up with the other 2 in separate PRs
* Suggestion: Add a test or two showing how this is intended to be used. Doesnt need to be comprehensive at this stage but will help put the code "in motion" making it easier to understand the user perspective and outputs

**Comment by mjhajharia** [comment]
hiii @canyon289 @fonnesbeck i'm so sorry i sort of went off the grid for half a week or so. will get everything done for the first evaluation before we meet on wednesday and then we can do finishing touches. meanwhile, might use slack for pytest queries, setting everything up rn

**Comment by mjhajharia** [comment]
> > so before any comments, this is a very very very rudimentary PR, right now i've only made it into a PR to be able to hear comments. My current plan for implementing sarima is:
> > ```
> > * [ ]  convert [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class
> > 
> > * [ ]  add aesara and suitable pymc functionalities to make it a pymc distribution (a separate submodule as @fonnesbeck suggested)
> > 
> > * [ ]  Add Kalman filter once we have a working sarima distribution in pymc3
> > 
> > * [ ]  additional improvements // features like a subclass of arma, ar, ma, sarimax + ipython notebooks with tutorials
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > so right now, i'm in the part where i've created a rudimentary conversion of [varstan/sarima.stan](https://github.com/asael697/varstan/blob/5378f428cad9560dae7f6daf8f431113f19a2019/inst/stan/Sarima.stan) into a working python class, its not working yet, the function definitions etc are all over the place, ive just essentially translated stan code to numpy/pymc equivalent. i could definitely definitely use opinions/suggestions/detailed comments/feedback anything!!!
> 
> Some comments
> 
> * This is a great draft PR. Thank you for sharing. If we ask a lot of questions its not because you did anything wrong, more so that this is the perfect place to ask precise questions and have great concrete discussion
> * Suggestion: In this PR I would focus on the first two items in the checklist, and the follow up with the other 2 in separate PRs
> * Suggestion: Add a test or two showing how this is intended to be used. Doesnt need to be comprehensive at this stage but will help put the code "in motion" making it easier to understand the user perspective and outputs

thanks!! yes i agree, i'll do the first two items before the first eval (15th july) and then get to the latter two in the next half of the summer. yes getting to testing today, i'm sorry about the procrastination, hit a depression week lol got nothing done, will get to it asap

File: pymc3/distributions/arimastan.py (hunk 2 above)
Comment (mjhajharia) [reply pymc3/distributions/arimastan.py]: yes pycharm pointed that out too, thinking i'll make separate utility functions

File: pymc3/distributions/arimastan.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        self.dd = dd  # number of seasonal differences
+        self.qq = qq  # number of predictions GARCH
+        self.period = period  # time series period
+        self.d1 = d1  # number of independent variables
+        self.sigma0 = sigma0  # Variance parameter
+        self.mu0 = mu0  # location parameter
+
+        # data
+
+        self.y = np.array(self.n1)  # outcome time series
```
Comment (MarcoGorelli) [inline pymc3/distributions/arimastan.py]: Is `self.n1` an integer? If so, this'll be zero-dimensional. If you want to make an array of `n1` zeros that you'll fill in later, you could do `np.zeros(self.n1)`

**Comment by asael697** [comment]
Hello I was invited to look the work, so

1) dont check varstan package anymore check bayesforecast package that has bugs fixed.

2) I can send a oversimplified stan code for the general SARiMAX model withouth the priors so you can see how it works and then you figure out how to include priors. 

 3)my stan code dont solve SARIMA(p,d,q), solve the ARMA(p,q) of the new series y1 = diff_d Y. Consider this when you are estimating the likelihood  and residuals

4) yes some models are simple cases of the general SARIMAX model.

5) for priors what I pass is a vector of dimension 4. Lets say you wanna put a Student_t prior with location 5, scale 4 degrees freedom 9 then the vector I send to Stan from R is (5,4,9,2) where the last digit 2 is the integer code of the used prior that is select with my ten thousands if_else. Check rstanarm package for a better idea(they crrate a stan function to deal with it more easy).

6) remember that arma models need to ve stationary to work so for an ARMA(1,1) the autorefressive parameter need to be between -1 and 1. So dont put big priors to it. Explosive processes in arima models are bad ideas.

7) sorry for my bad english and nice project.

**Comment by mjhajharia** [comment]
> 2\. I can send a oversimplified stan code for the general SARiMAX model with puth the priors so you can ser how it works and then you figure put how to include priors.

hi! yes that would be wonderful, thankyou for taking the time to reply, appreciate it!!!



> 2\. for priors what I pass is a vector of dimension 4. Lets say you wanna put a Student_t prior with location 5, scale 4 degrees freedom 9 then the vector I send to Stan from R is (5,4,9,2) where the last digit 2 is the integer code of the used prior that is select with my ten thousands if_else. Check rstanarm package for a better idea(they crrate a stan function to deal with it more easy).

okay so for this um, I'll not be creating vectors i think in pymc3 i will be creating distributions and treat those parameters as random variables. i saw yeah sarima.R had vectors, i replaced all of the stan distributions with pymc3, lots of stuff still needs to be worked out

> 3\. remember that arma models need to ve stationary to work so for an ARMA(1,1) the autorefressive parameter need to be between -1 and 1. So dont put big priors to it. Explosive processes in arima models are bad ideas.

yes! i plan to get to the stationarity thing eventually, for now we'll assume data is already differenced (D=0) if xreg is None, i think that's what happens in sarima.R right?



> 4\. sorry for my bad english and nice project.

don't worry about that, thankyou!!!

**Comment by asael697** [comment]
2) yes I just explain the vector stuff to maoe my stan code easiser to understand.

3) yes I dont aloud seasonal differences if there are regressor variables gets to complicated to track and eventually you need to inverse difference all (integratr).

Also ARMA models are stochastic process (set of rabdom vatiables) so they can not be distributions.

**Comment by mjhajharia** [comment]
> 1. yes I just explain the vector stuff to maoe my stan code easiser to understand.
> 2. yes I dont aloud seasonal differences if there are regressor variables gets to complicated to track and eventually you need to inverse difference all (integratr).
> 
> Also ARMA models are stochastic process (set of rabdom vatiables) so they can not be distributions.

my bad, like priors can be distributions right

**Comment by asael697** [comment]
A prior is a distribution for modelating the parametric space of an arbitrary parameter. 

And stochastic process can be use to "build priors" in a Hilbert probability space just like GPs and Dirichlet processes. But saddly ARMA processes are npt robust enough to be used as functional priors. 

So yes priors are distributions but not all stochastic processes can be used to build priors

**Pull request closed without merging**